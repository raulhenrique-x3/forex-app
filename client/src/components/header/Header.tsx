import {
  Avatar,
  Box,
  Button,
  Container,
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerFooter,
  DrawerHeader,
  DrawerOverlay,
  Flex,
  Text,
  useDisclosure,
  WrapItem,
} from "@chakra-ui/react";
import React from "react";
import { HiOutlineMenu } from "react-icons/hi";
import styles from "./header.module.scss";
import useAuth from "../../hooks/useAuth";
import { Link, useNavigate, useParams } from "react-router-dom";
import { DesktopHeaderInfos } from "../desktopHeaderInfos/DesktopHeaderInfos";
import { IUser } from "../../interface/interface";
import { Wallet } from "../wallet/Wallet";

interface IHeader {
  user: IUser[];
}

const Header: React.FC<IHeader> = ({ user }) => {
  const navigate = useNavigate();
  const { isAuthenticated, logout } = useAuth();
  let { userId } = useParams();

  const userLogout = () => {
    logout();
    if (!isAuthenticated) {
      navigate("/");
    } else {
      return;
    }
  };

  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <header className={styles.header}>
      <Link to={`/home/${userId}`}>
        <p className={styles.headerText}>Forex-app</p>
      </Link>
      <div className={styles.desktopView}>
        <Button rounded={"full"} bgColor={"#ffffff"} onClick={onOpen}>
          <HiOutlineMenu color="#8236FD" />
        </Button>
        <Drawer isOpen={isOpen} placement="right" onClose={onClose}>
          <DrawerOverlay />
          <DrawerContent backgroundColor={"#F1F1F1"}>
            <DrawerCloseButton />

            <DrawerBody display={"flex"} flexDirection={"column"} alignItems={"center"}>
              <Box
                marginTop={8}
                width="100%"
                backgroundColor="#8236fd"
                color="#000000"
                display={"flex"}
                flexDirection={"row"}
                alignItems={"center"}
                gap={1}
                borderRadius={5}
                padding={2}
                marginBottom={4}
              >
                <Avatar backgroundColor={"#F1F1F1"} color={"#000000"} size="lg" name={user[0]?.userName} />
                <Container>
                  <Text fontSize="lg" color={"#ffffff"} fontWeight="bold">
                    {user[0]?.userName}
                  </Text>
                  <Text fontSize="md" color={"#ffffff"}>
                    {user[0]?.userEmail}
                  </Text>
                </Container>
              </Box>
              <Container>
                <Wallet />
              </Container>
              <Container marginTop={8}>
                <Flex direction={"column"} gap={8}>
                  <Text fontWeight={600} color={"#8236fd"} borderBottom={"1px solid #8236fd"}>
                    <Link to={`/home/${userId}`}>Home</Link>
                  </Text>
                  <Text fontWeight={600} color={"#8236fd"} borderBottom={"1px solid #8236fd"}>
                    <Link to={`/wallet/${userId}`}>Wallet </Link>
                  </Text>
                  <Text fontWeight={600} color={"#8236fd"} borderBottom={"1px solid #8236fd"}>
                    <Link to={`/history/${userId}`}>History </Link>
                  </Text>
                </Flex>
              </Container>
            </DrawerBody>

            <DrawerFooter>
              <Button colorScheme="blue" onClick={userLogout}>
                Logout
              </Button>
            </DrawerFooter>
          </DrawerContent>
        </Drawer>
      </div>

      <nav className={styles.desktopNav}>
        <ul className={styles.desktopUl}>
          <li>
            <Link to={`/home/${userId}`}>Home </Link>
          </li>

          <li>
            <Link to={`/wallet/${userId}`}>Wallet </Link>
          </li>
          <li>
            <Link to={`/history/${userId}`}>History </Link>
          </li>
        </ul>
        <Button colorScheme="blue" onClick={userLogout}>
          Logout
        </Button>
      </nav>
    </header>
  );
};

export default Header;
