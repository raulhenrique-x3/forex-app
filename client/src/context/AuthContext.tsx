import React, { createContext, useState } from "react";

interface IProps {
  isAuthenticated: boolean;
  login: (value: boolean) => void;
  logout: () => void;
}

export const AuthContext = createContext<IProps>({
  isAuthenticated: false,
  login() {},
  logout() {},
});

interface IChildren {
  children: React.ReactNode;
}

export function AuthContextProvider({ children }: IChildren) {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(false);

  const login = (value: boolean) => {
    setIsAuthenticated(value);
  };

  const logout = () => {
    setIsAuthenticated(false);
  };

  return <AuthContext.Provider value={{ isAuthenticated, login, logout }}>{children}</AuthContext.Provider>;
}

export const AuthConsumer = AuthContext.Consumer;
