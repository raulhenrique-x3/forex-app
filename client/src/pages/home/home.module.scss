.desktopViewInfo {
  display: none;
}

@media (min-width: 768px) {
  .desktopViewInfo {
    display: block;
    width: 100%;
  }
}
