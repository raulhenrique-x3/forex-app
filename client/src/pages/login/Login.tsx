import {
  Box,
  Button,
  Card,
  CardBody,
  CardHeader,
  Container,
  FormLabel,
  Input,
  Stack,
  StackDivider,
  Text,
  useToast,
} from "@chakra-ui/react";
import useAuth from "../../hooks/useAuth";
import { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import axios from "axios";

interface ILogin {
  submit: ({ userEmail, userPassword }: any) => void | undefined;
}

export const Login: React.FC<ILogin> = ({ submit }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const toast = useToast();
  const { login } = useAuth();
  const navigate = useNavigate();

  const userLogin = () => {
    submit({ userEmail: "raul@email.com", userPassword: "123456789" });

    axios
      .post("http://localhost:5000/session/", { userEmail: email, userPassword: password })
      .then((res) => {
        login(true);
        toast({
          title: "Logged",
          status: "success",
          isClosable: true,
        });
        navigate(`/home/${res.data._id}`);
      })
      .catch((err) => {
        toast({
          title: err.response.data.message,
          status: "error",
          isClosable: true,
        });
      });
  };

  return (
    <Box
      backgroundColor={"#cccccc"}
      width={"100vw"}
      height={"100vh"}
      display="flex"
      alignItems="center"
      justifyContent="center"
      flexDirection="column"
    >
      <Container>
        <Card>
          <CardHeader>
            <Text fontSize={"1.5rem"} fontWeight={"bold"}>
              Login
            </Text>
          </CardHeader>
          <CardBody>
            <Stack divider={<StackDivider />} spacing="4">
              <Box>
                <FormLabel>Email</FormLabel>
                <Input
                  data-testid="email"
                  type="text"
                  placeholder="Enter your email"
                  onChange={(e) => setEmail(e.target.value)}
                />
              </Box>
              <Box>
                <Text>Password</Text>
                <Input
                  type="password"
                  data-testid="password"
                  placeholder="Enter your password"
                  onChange={(e) => setPassword(e.target.value)}
                />
              </Box>
              <Box display={"flex"} flexDirection={"column"}>
                <Button
                  data-testid="loginButton"
                  color={"#ffffff"}
                  mt={4}
                  backgroundColor={"#8236FD"}
                  onClick={userLogin}
                >
                  Login
                </Button>
                <Link to="/register">
                  <Container marginTop={4}>
                    <Text
                      textAlign={"center"}
                      fontWeight={600}
                      color={"#8236FD"}
                      _hover={{ borderBottom: "1px solid #8236FD" }}
                    >
                      Create an account
                    </Text>
                  </Container>
                </Link>
              </Box>
            </Stack>
          </CardBody>
        </Card>
      </Container>
    </Box>
  );
};
