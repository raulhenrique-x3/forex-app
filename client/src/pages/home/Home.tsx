import { Box, Flex, Divider, Container } from "@chakra-ui/react";
import axios from "axios";
import styles from "./home.module.scss";
import { useState, useEffect } from "react";
import { useParams } from "react-router";
import CurrencyContainer from "../../components/currencyContainer/CurrencyContainer";
import { DesktopHeaderInfos } from "../../components/desktopHeaderInfos/DesktopHeaderInfos";
import Header from "../../components/header/Header";
import { Wallet } from "../../components/wallet/Wallet";
import { IUser } from "../../interface/interface";
import io from "socket.io-client";

export const Home = () => {
  const socket = io("ws://localhost:5000", { autoConnect: true });
  const [userData, setUserData] = useState<IUser[]>([]);
  const [ioResponse, setIoResponse] = useState<object[]>();
  let { userId } = useParams();

  useEffect(() => {
    socket.on("Atualized data from API", (data) => {
      setIoResponse([data]);
    });

    return () => {
      socket.off("Atualized data from API");
    };
  }, [socket]);

  useEffect(() => {
    axios
      .get(`http://localhost:5000/home/${userId}`)
      .then((res) => {
        setUserData([res.data]);
      })
      .catch((error) => {
        console.error(error);
      });
  }, [ioResponse]); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <Flex w="100%" direction="column" alignItems="center" justifyContent="center" backgroundColor={"#F1F1F1"}>
      <Header user={userData} />
      <div className={styles.desktopViewInfo}>
        <DesktopHeaderInfos user={userData} />
      </div>
      <Container w="100%" height={"100vh"} maxWidth="400px" p={4}>
        <Box pt={4}>
          <Wallet cursor="pointer" showArrow={true} />
        </Box>
        <Divider />
        <Box>
          <CurrencyContainer showArrow={true} showGBP={true} showUSD={true} userId={userId} />
        </Box>
        <Divider />
      </Container>
    </Flex>
  );
};
