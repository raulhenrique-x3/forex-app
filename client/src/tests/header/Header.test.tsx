import { Route, Routes, createMemoryRouter, RouterProvider, BrowserRouter } from "react-router-dom";
import { AuthContextProvider } from "../../context/AuthContext";
import { cleanup, render, screen, act, fireEvent, waitFor } from "@testing-library/react";
import "setimmediate";
import Header from "../../components/header/Header";
import { Login } from "../../pages/login/Login";
import { Home } from "../../pages/home/Home";
import { enableFetchMocks } from "jest-fetch-mock";

enableFetchMocks();
jest.setTimeout(60000);
afterAll(cleanup);

describe("He tries to logoff", () => {
  it("should be click in the Logout button", async () => {
    const logoutEvent = { logout: "test" };
    const routes = [
      {
        path: "/home/63e5135202704aea76e5fc52",
        element: <Home />,
        loader: () => logoutEvent,
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ["/", "/home/63e5135202704aea76e5fc52"],
      initialIndex: 1,
    });

    act(() => {
      render(
        <AuthContextProvider>
          <BrowserRouter>
            <Header user={[]} />
            <Routes>
              <Route path="/" element={<Login submit={() => {}} />} />
            </Routes>
          </BrowserRouter>
          <RouterProvider router={router} />
        </AuthContextProvider>
      );
    });
    const logoutButton = screen.getByTestId("Logout");
    expect(logoutButton).toBeTruthy();
    fireEvent.click(logoutButton);
    await waitFor(() => screen.queryByPlaceholderText("Enter your email"));
    expect(screen.queryByPlaceholderText("Enter your email")).toBeTruthy();
  });
});
