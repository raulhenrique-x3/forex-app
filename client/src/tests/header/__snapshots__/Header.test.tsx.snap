// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`He tries to logoff should be click in the Logout button 1`] = `
<button
  class="chakra-button css-4xx2wk"
  data-testid="Logout"
  type="button"
>
  Logout
</button>
`;
